src/Firestore/DocumentSnapshot.php

```diff
@@ -0,0 +1,314 @@
+    }
+
+    /**
+     * Returns document field data as an array.
+     *
+     * Example:
+     * ```
+     * $fields = $snapshot->fields();
+     * ```
+     *
+     * @return array
+     */
+    public function fields()
+    {
+        return $this->fields;
+    }
+
+    /**
+     * Returns true if the document exists in the database.
+     *
+     * Example:
+     * ```
+     * if ($snapshot->exists()) {
+     *     echo "The document exists!";
+     * }
+     * ```
+     *
+     * @return bool
+     */
+    public function exists()
+    {
+        return $this->exists;
+    }
+
+    /**
+     * Get a field by field path.
+     *
+     * A field path is a string containing the path to a specific field, at the
+     * top level or nested, delimited by `.`. For instance, the value `hello` in
+     * the structured field `{ "foo" : { "bar" : "hello" }}` would be accessible
+     * using a field path of `foo.bar`.
+     *
+     * Example:
+     * ```
+     * $value = $snapshot->get('wallet.cryptoCurrency.bitcoin');
+     * ```
+     *
+     * ```
+     * // Field names containing dots or symbols can be targeted using a FieldPath instance:
+     * use Google\Cloud\Firestore\FieldPath;
+     *
+     * $value = $snapshot->get(new FieldPath(['wallet', 'cryptoCurrency', 'my.coin']));
+     * ```
+     *
+     * @param string|FieldPath $fieldPath The field path to return.
+     * @return mixed
+     * @throws \InvalidArgumentException if the field path does not exist.
+     */
+    public function get($fieldPath)
+    {
+        $res = null;
+
+        if (is_string($fieldPath)) {
+            $parts = explode('.', $fieldPath);
+        } elseif ($fieldPath instanceof FieldPath) {
+            $parts = $fieldPath->path();
+        } else {
+            throw new \InvalidArgumentException('Given path was not a string or instance of FieldPath.');
+        }
+
+        $len = count($parts);
+
+        $fields = $this->fields;
+        foreach ($parts as $idx => $part) {
+            if ($idx === $len-1 && isset($fields[$part])) {
+                $res = $fields[$part];
+                break;
+            } else {
+                if (!isset($fields[$part])) {
+                    throw new \InvalidArgumentException('field path does not exist.');
+                }
+
+                $fields = $fields[$part];
+            }
+        }
+
+        return $res;
+    }
+
+    /**
+     * @access private
+     */
+    public function offsetSet($offset, $value)
+    {
+        throw new \BadMethodCallException('DocumentSnapshots are read-only.');
+    }
+
+    /**
+     * @access private
+     */
+    public function offsetExists($offset)
+    {
+        return isset($this->fields[$offset]);
+    }
+
+    /**
+     * @access private
+     */
+    public function offsetUnset($offset)
+    {
+        throw new \BadMethodCallException('DocumentSnapshots are read-only.');
+    }
+
+    /**
+     * @access private
+     */
+    public function offsetGet($offset)
```

Comment:
The Objective-C client does something similar and supports brackets for field retrieval (e.g. data[field]). Because of this, we decided to not offer a .get() method. It might make sense to consider this for PHP as well.

Reply:
I like having `get()` to support FieldPaths. array-style access won't support that.

Reply:
Thanks for the explanation.